### PR TITLE
More math fns

### PR DESCRIPTION
**Pull request by plynchnlm**

### CODE REVIEW

File: test/cases/math.yaml [hunk 1]
```diff
@@ -11,8 +11,12 @@ tests:
   - desc: Error adding non-number
     expression: MathTestData.n1 + MathTestData.s5
     error: true
-
-
+  - desc: Can subtract two numbers
+    expression: MathTestData.n1 - MathTestData.n2
+    result: [-3]
```
Comment (akanduru) [inline test/cases/math.yaml]: Going by 6.6.3 and 6.6.7, the string concatination should include 
'a' + 'b' = 'ab'
'a' & 'b' = 'ab'
'a' & '' = 'a'
'a' + '' = '' 
The code here seems to violate the last statement, right?

File: src/fhirpath.js [hunk 2]
```diff
@@ -395,11 +395,27 @@ function AdditiveExpression(ctx, parentData, node) {
     throw "AdditiveExpression:  Was expecting one element but got " +JSON.stringify(right);
   left = leftIsArray ? left[0] : left;
   right = rightIsArray ? right[0] : right;
-  if (Number.isNaN(left))
-    throw "AdditiveExpression:  Was expecting a number but got " +JSON.stringify(left);
-  if (Number.isNaN(right))
-    throw "AdditiveExpression:  Was expecting a number but got " +JSON.stringify(right);
-  return [left + right];
+  let operator = node.terminalNodeText[0];
```
Comment (akanduru) [inline src/fhirpath.js]: I was thinking all math numbers only operations in 6.6 need the same checks above. We could perhaps move these checks into a separate function and use them for those operations.

File: src/parser/index.js [hunk 3]
```diff
@@ -57,6 +57,14 @@ var parse = function(path){
           parentNode.children = [];
         parentNode.children.push(node);
         parentStack.push(node);
+        // Also collect this node's terminal nodes, if any.  Terminal nodes are
+        // not walked with the rest of the tree, but include things like "+" and
+        // "-", which we need.
+        node.terminalNodeText = [];
+        for (let c of ctx.children) {
```
[the remark this answers is not in the document]
Comment (plynchnlm) [reply src/parser/index.js]: Webpack can transpile for us.  However, it isn't configured to do that currently, so I have opened a new issue (#23).

File: src/fhirpath.js (hunk 2 above)
Comment (plynchnlm) [reply src/fhirpath.js]: Possibly.  If that is the case, I will refactor those into a separate function when the other functionI(s) that need it are written.

File: test/cases/math.yaml (hunk 1 above)
Comment (plynchnlm) [reply test/cases/math.yaml]: "Empty" means "empty collection", i.e., [ ] not ''.  However, you correctly point out that + should be able to concatenate strings, which I had missed.

Comment (plynchnlm) [reply test/cases/math.yaml]: Fixed.

File: src/fhirpath.js [hunk 4]
```diff
@@ -405,16 +403,25 @@ function AdditiveExpression(ctx, parentData, node) {
...
-      throw "AdditiveExpression: Unexpected operator: " +operator;
+    if (left.length === 0 || right.length === 0)
+      return [];
+    else {
+      left = left[0];
+      right = right[0];
+      if (typeof left !== typeof right)
+        throw "AdditiveExpression:  Mismatched types, "+typeof left +" and "+ typeof right;
+      if (operator === "+")
+        return [left + right];
```
Comment (akanduru) [inline src/fhirpath.js]: This may fail 'a' + [] = [] (spec 6.6.7). I don't see a test for it.

Comment (plynchnlm) [reply src/fhirpath.js]: That would be the test cases "Empty result if one or both operands is empty".  I added a new test for the case where the non-empty one is a string.

**Pull request merged**